### DISCUSSION TITLE
Bumble backend?

### DISCUSSION THREAD

**Discussion by JPHutchins** [hbldh/bleak#1588] [Ideas]
In a move that amuses me to no end, Google developed a Bluetooth stack that sidesteps OS implementations: https://github.com/google/bumble

I haven't played with it yet, and suspect that it's not particularly user friendly.  Specifically, "rather wacky" steps need to be taken to uninstall the OS USB drivers for BT adpaters so that Bumble can talk to them instead: https://google.github.io/bumble/platforms/index.html

That said, an increase in cross-platform stability may be well worth some one-time setup headache for the many researchers and developers who use Bleak, if not end users.  Users probably want their headphones and keyboard to keep working 🤣.

**Comment by JPHutchins** [comment]
Could be a nice swiss-army knife!

https://google.github.io/bumble/platforms/zephyr.html

https://www.nordicsemi.com/Products/Development-hardware/nRF52840-Dongle

The big advantage being that users can update their dongle FW with what they need and get good insights into problems.

**Comment by JPHutchins** [comment]
Seems asyncio ready though I'm a bit skeptical of this AsyncRunner abstraction that's been created.

They also use `serial_asyncio` which is simply a python thread hack of PySerial.  POSIX and Windows have proper async support for serial and I'm always saying that I'll "get around to replacing pyserial" but it's a lot of work to handle MacOS, Windows, and Linux.  And it's a meaningless improvement with how fast the host computer is compared to the serial device.  I suppose that the same is true for BLE.  I just think threads are silly 😛.  https://github.com/google/bumble/blob/090309302f77c3f992e5c4a4f92269da90244cdb/bumble/transport/serial.py#L20

Perhaps the USB transport uses async?

**Comment by vChavezB** [comment]
Hi @JPHutchins ,

I have added initial support for Bumble in this PR https://github.com/hbldh/bleak/pull/1681

It is currently an experimental feature I am testing but I am looking forward for some feedback to improve it.

I wrote some small instructions over [here](https://github.com/vChavezB/bleak/blob/bumble/docs/backends/bumble.rst). But basically you can set the transport layer for bumble by adding the `adapter` argument to the `BleakScanner` or `BleakClient`. As Bleak selects the backend based on the OS, the env. variable `BLEAK_BUMBLE ` must be set to select this backend.
